pam/pam_test/module-transaction-dummy_test.go

```diff
@@ -0,0 +1,512 @@
+package pam_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/msteinert/pam"
+	"github.com/stretchr/testify/require"
+)
+
+func ptrValue[T any](value T) *T {
+	v := value
+	return &v
+}
+
+func TestSetGetItem(t *testing.T) {
+	t.Parallel()
+	t.Cleanup(MaybeDoLeakCheck)
+	t.Cleanup(MaybeDoLeakCheck)
+
+	testCases := map[string]struct {
+		item             pam.Item
+		value            *string
+		expectedGetError error
+		expectedSetError error
+		expectedValue    *string
+	}{
+		"set invalid item": {
+			item:             pam.Item(-1),
+			value:            ptrValue("some value"),
+			expectedSetError: pam.BadItem,
+		},
+
+		"get invalid item": {
+			item:             pam.Item(0),
+			expectedGetError: pam.BadItem,
+			expectedValue:    ptrValue(""),
+		},
+
+		"set user": {
+			item:  pam.User,
+			value: ptrValue("an user"),
+		},
+
+		"get unset user": {
+			item:          pam.User,
+			expectedValue: ptrValue(""),
+		},
+
+		"get and set user": {
+			item:          pam.User,
+			value:         ptrValue("the-user"),
+			expectedValue: ptrValue("the-user"),
+		},
+	}
+
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Cleanup(MaybeDoLeakCheck)
+
+			require.True(t, tc.value != nil || tc.expectedValue != nil)
```

Comment:
We control the values in the test case, so we don't need this assertion here, IMHO

Reply:
True, but it's like an assertion here... To avoid programming errors. Not sure if there's a better way than use `require` here.

One may end up not setting these values and not test anything at all.

Reply:
I've changed to use `Truef` though to explain why.

Reply:
I can see the point of the assertion, I just wonder whether trying to shield it from programming errors is necessary. They are prone to happen, indeed, but if they happen either the tests will fail or the program will fail to build so we don't really need this assertion, IMHO.
Also, I also think that submitting `nil` values to these functions can actually be test cases, no? Like: what do we expect to happen if we try to `tx.SetItem()` with a nil value? I know this sounds "trivial", but those test cases are very important as well.

Reply:
Instead of asserting, you can also use something like:
```Go
// These test inputs can not be nil
if tc.value != || tc.expectedValue != nil {
    t.Fatalf("Setup: some test inputs were nil, but they must not be")
}
```
This is what `require` calls, basically. I'm fine with using require for the assertions, though. But I also think that those could be test cases (error ones).

Reply:
Indeed we do test the case were we have empty values (if you see the tests, I always try to focus more on the edge cases since those are the ones that are more often causing troubles), but in this case I want to ensure that we're actually testing something.

I.e. having an empty value as input is fine, but in that case we should provide an expected value too. In fact that's a pointer for a reason, if it would have been a string then it wouldn't have been possible to test the `""` case, but if that's null we can't really compare it with a function that will always return a `string` and not a `*string`.

Reply:
So the check would actually be:

```go
// These test inputs can not be all unset
if tc.value == "" && tc.expectedValue == nil {
    t.Fatalf("Setup: some test inputs were nil, but they must not be")
}
```